horace_core/sqw/@Experiment/private/build_from_old_headers_.m

```diff
@@ -1,37 +1,62 @@
+%   >> headers == varargin{1}: cell array of old-style header structs
+%
+% Output
+% -------
+%   >> obj: an Experiment object as the new-style header packaging the
+%           old-style header info
+%
+% COMMENT:
+% This should really be static method - it relates to Experiment but does
+% not use Experiment object contents
+%------------------------------------------------------------------------
+
+% Name varargin{1} as (old-style) headers to clarify what is coming in
 headers = varargin{1};
-expdata = repmat(IX_experiment,1,numel(headers));
-instruments = repmat(IX_inst(),1,numel(headers));
-samples = repmat(IX_samp(),1,numel(headers));
 
-% convert old headers, resored differently from sqw and mat files into the
-% same format
+% Make arrays of the hew-style header class types to receive the data
+% coming from each run in headers
+expdata     = repmat(IX_experiment(), 1,numel(headers));
+instruments = repmat(IX_null_inst(),  1,numel(headers));
+samples     = repmat(IX_null_sample(),1,numel(headers));
+
+% convert old headers, restored differently from sqw and mat files into the
+% same format - there appear to be cases where old-style header structs and
+% Experiments mix - this converts to old-style before we start, though not
+% not convinced it is needed or correct
 headers = normalize_old_headers(headers);
-%
+
+% Now operate on the assumed old-style header structs, either received or
+% converted
 for i=1:numel(headers)
+    % Name  i'th header locally
     hdr = headers{i};
+    
+    % Unpack lattice parameters for putting into sample
     alatt = hdr.alatt;
     angdeg = hdr.angdeg;
+    
+    % Local name the instrument (not liking plural instruments - 
+    % should be the single instrument from the old-syle header structure
     instr = hdr.instruments;
     
+    % If instr is a struct
     if isstruct(instr)
+        % Instrument in header is empty struct, replace with null instrument
         if isempty(instr) || numel(fieldnames(instr))==0
-            % do nothing, its empty instrument and its already there
-        elseif isfield(instr,'fermi_chopper') %TODO: should it be instrument factory which produces instruments as function of input parameters?
-            % This would allow easy modify and add new instruments, not to
-            % run through code looking for places to identify them
-            instrument = IX_inst_DGfermi(instr.moderator, ...
-                instr.aperture,  ...
-                instr.fermi_chopper);
-            instruments(i) = instrument;
+            instruments(i) = IX_null_inst();
+        % Struct may have data to make an instrument, delegate to factory
+        % method (which may reject the struct and throw)
         else
-            error('HORACE:Experiment:invalid_argument',...
-                'unknown structure tried to be recovered as instrument');
+            instruments(i) = make_instrument_from_struct(instr);
         end
```

Comment:
dont we have instrument constructor for this?

Reply:
We have constructors for the individual instruments, null, DGdisk, DGfermi. Here we have an unknown instrument defined through a struct, and need the function to return the correct instrument class object. Even with proper polymorphic behaviour in Matlab (not present) and handle classes (we're not using them and I don't think they provide anything here) the discovery of the struct type has not be done by something that is not a constructor before the construction can be done.